src/layers/domain/core/questionnaire.py

```diff
@@ -1,75 +1,149 @@
-from datetime import datetime
-from enum import Enum
-from uuid import UUID
+from datetime import date, datetime, time
+from types import FunctionType
+from typing import Generic, Type, TypeVar
 
-from pydantic import BaseModel, Field
+from domain.core.error import DuplicateError, InvalidResponseError
+from domain.core.validation import ENTITY_NAME_REGEX
+from pydantic import BaseModel, Field, validator
 
-from .error import DuplicateError
-from .validation import ENTITY_NAME_REGEX
+# from domain.core.questionnaire_validation_custom_rules import url
 
-
-class QuestionType(Enum):
-    """
-    The data type of the question
-    """
-
-    STRING = str
-    INT = int
-    BOOL = bool
-    DATE_TIME = datetime
+T = TypeVar("T")
+A = {str, int, bool, datetime, float, date, time}  # allowed question types
 
 
-class Question(BaseModel):
+class Question(BaseModel, Generic[T]):
     """
     A single Questionnaire Question
     """
 
+    class Config:
+        """
+        Validation rules are not pydantic classes
+        """
+
+        arbitrary_types_allowed = True
+
     name: str
-    type: QuestionType
+    type: T
     multiple: bool
+    validation_rules: set[FunctionType] = None
+    choices: set[T] = None
+
+    @validator("type")
+    def validate_question_type(cls, type):
+        if type not in A:
+            raise ValueError
+        return type
 
 
 class Questionnaire(BaseModel):
     """
     A Questionnaire represents a collection of Questions, in a specific order.
     """
 
-    id: UUID
     name: str = Field(regex=ENTITY_NAME_REGEX)
-    _questions: list[Question] = (lambda: [])()
+    version: int
+    questions: dict[str, Question] = Field(default_factory=dict)
 
     def __contains__(self, question_name: str) -> bool:
         """
         Returns true if the question specified exists within the questionnaire
         """
-        return question_name in (q.name for q in self._questions)
+
+        return question_name in self.questions
 
     def add_question(
         self,
         name: str,
-        type: QuestionType = QuestionType.STRING,
+        type: Type = str,
```

Comment:
so a potential bug that might come up is that the word `type` also refers a python global called `type` (i.e. you can do `type("hello")` which returns `str`). It's actually quite an important object in python, so it might be worth renaming anything in this file called `type` to `answer_type` to avoid weird side-effects coming up at a later date